File: python/person.py
class Person:
    USER_ROLE_ADMIN = 0
    USER_ROLE_ENGINEER = 1
    USER_ROLE_MANAGER = 2
    USER_ROLE_SALES = 3

    def __init__(self, role: int, swedish_personal_number: str, phone_number: str):
        self.role = role
        self.set_role(role)
        self.swedish_personal_number = swedish_personal_number
        self.set_swedish_personal_number(swedish_personal_number)
        self.phone_number = phone_number

    def get_role(self) -> int:
        return self.role

    def set_role(self, role):
        if role < 0 or role > 4:
            raise ValueError("illegal role " + role)
        self.role = role

    def get_swedish_personal_number(self) -> str:
        return self.swedish_personal_number

    def set_swedish_personal_number(self, swedish_personal_number):
        swedish_personal_number = swedish_personal_number.replace("-", "")
        if swedish_personal_number.__len__() != 12:
            raise ValueError("invalid personal number " + swedish_personal_number)
        self.swedish_personal_number = swedish_personal_number

    def get_phone_number(self) -> str:
        return self.phone_number

    def set_phone_number(self, phone_number: str):
        self.phone_number = phone_number

    def birth_year(self) -> int:
        year = self.swedish_personal_number[0:4]
        return int(year)

    def country_code(self) -> str:
        code = ""
        if self.phone_number.startswith("00"):
            code = self.phone_number[2:4]
        elif self.phone_number.startswith("+"):
            code = self.phone_number[1:3]
        if code != "":
            return "+" + code
        return ""

    def can_delete_users(self) -> bool:
        return self.role == Person.USER_ROLE_MANAGER or self.role == Person.USER_ROLE_ADMIN

File: kotlin/src/main/kotlin/Person.kt
package sammancoaching

class Person(
    role: Int,
    swedishPersonalNumber: String,
    phoneNumber: String
) {
    private var role = 0
    var swedishPersonalNumber: String? = null
        private set
    var phoneNumber: String

    init {
        setRole(role)
        setSwedishPersonalNumber(swedishPersonalNumber)
        this.phoneNumber = phoneNumber
    }


    fun setRole(role: Int) {
        require(!(role < 0 || role > 4)) { "illegal role$role" }
        this.role = role
    }

    fun setSwedishPersonalNumber(swedishPersonalNumber: String) {
        var swedishPersonalNumber = swedishPersonalNumber
        swedishPersonalNumber = swedishPersonalNumber.replace("-", "")
        require(swedishPersonalNumber.length == 12) { "invalid personal number $swedishPersonalNumber" }
        this.swedishPersonalNumber = swedishPersonalNumber
    }

    fun birthYear(): Int {
        val year = swedishPersonalNumber!!.substring(0, 4)
        return year.toInt()
    }

    fun countryCode(): String {
        var code = ""
        if (phoneNumber.startsWith("00")) {
            code = phoneNumber.substring(2, 4)
        }
        else if (phoneNumber.startsWith("+")) {
            code = phoneNumber.substring(1, 3)
        }

        if (code.isNotEmpty()) return  "+$code"
        return ""
    }

    fun canDeleteUsers(): Boolean {
        return role == USER_ROLE_MANAGER || role == USER_ROLE_ADMIN
    }

    companion object {
        const val USER_ROLE_ADMIN = 0
        const val USER_ROLE_ENGINEER = 1
        const val USER_ROLE_MANAGER = 2
        const val USER_ROLE_SALES = 3
    }
}
